app/assets/javascripts/controllers/ems_keypair/ems_keypair_controller.js

```diff
@@ -63,9 +63,10 @@
   };
 
   EmsKeypairController.prototype.showValidate = function(tab) {
-    var openstackAndNew = (this.model.emstype === 'openstack_infra') && this.newRecord;
+    var openstackInfraAndNew = (this.model.emstype === 'openstack_infra') && this.newRecord;
+    var openstackCloud = this.model.emstype === 'openstack';
```

Comment:
Is cloud supposed to be different from openstack infra and rhevm here?

Why is validation different for non-new openstack records than for non-new openstack infra? (Any chance that's a bug?)

Reply:
Cloud is just like rhevm here -- no validation for either 'new' or 'update' forms. Cloud is not like openstack infra because for OpenStack infra, we only validate for 'update', not for 'new'. If you look at the conditional before, we treat rhevm and cloud the same. The distinction is made for OpenStack Infra -- where we validate for update form but not new form. It's not a bug -- it's the way it's supposed to work, since we don't necessarily have an instance running at this point in which we can validate the keypair credentials -- this is the case for both rhevm and openstack cloud.